Fix long lines sent to editor

rybarczykj opened the pull request:
This is a working fix to #727, where long lines sent to the editor (usually with F7) are cut off and can't be edited correctly.

From what I can tell, the solution seems to work fine, but since this is the most code I've written at once for bpython, I want to make sure the structure / conventions / methods are valid. So I'll start it as a draft.

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -395,9 +396,9 @@ def __init__(
-        # lines separated whenever logical line
-        # length goes over what the terminal width
-        # was at the time of original output
+        # this is every line that's been displayed (input and output).
+        #   seperated logically, but if a line exceeded the terminal width
+        #   at the time of output, it is split into multiple entries
```
```suggestion
        # this is every line that's been displayed (input and output)
        # as with formatting applied. Logical lines exceeded the terminal width
        # at the time of output are split across multiple entries in this list.
```

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -411,7 +412,8 @@ def __init__(
         self.scroll_offset = 0
 
-        # from the left, 0 means first char
+        # cursor position relative to start of current_line
+        #   0 means first char
```
```suggestion
        # cursor position relative to start of current_line, 0 is first char
```

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -994,15 +996,58 @@ def send_current_block_to_external_editor(self, filename=None):
 
     def send_session_to_external_editor(self, filename=None):
+        """
+        Sends entire bpython session to external editor to be edited. Usually called with F7. 
```
```suggestion
        Sends entire bpython session to external editor to be edited. Usually bound to F7. 
```
to me "called with" means the arguments passed in

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -994,15 +996,58 @@ def send_current_block_to_external_editor(self, filename=None):
+        """
+        Sends entire bpython session to external editor to be edited. Usually called with F7. 
+
+        Loops through display_lines and writes lines to editor. Whenever a line starts with ps1 or ps2, 
+        we actually pop and write from history instead of display_lines. This is because display_lines 
+        cuts off long lines but self.history preserves them.
```
ooo tricky!

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -1903,6 +1948,11 @@ def initialize_interp(self):
         del self.coderunner.interp.locals["_Helper"]
 
     def getstdout(self):
```
Heads up that this method is also called by write2file(), copy2clipboard(), and pastebin() in bpython/repl.py so we may or may not want to change its behavior (depends on whether the changes make sense for those methods too)

rybarczykj replied:
Oh whoops, I accidentally left that outdated TODO in there. Yeah I agree, we shouldn't need to change this method.

rybarczykj commented:
Ok as Tom and I discussed (outside of this thread), I've added a new list, `self.all_logical_lines` which keeps track of all input and output lines without wrapping or formatting. The data structure is a list of tuples, where the first element is a string of the line and the second element is one of 2 strings: "input" or "output". From all the tests and checks I ran, I'm pretty sure I've added it in all the correct spots so that it updates and undoes correctly.

This makes it very easy to write to editor without having to worry about wrapping.

Also, as we talked about, possible future extensions of this would be making `self.history` dependent on this list, to save space (`self.history` is always equal to all "input" lines in `self.all_logical_line`). This list could be useful to use with copying and pasting code blocks, where wrapping is so commonly an issue.

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -1870,8 +1868,8 @@ def prompt_for_undo():
             temp = self.redo_stack.pop()
-            self.push(temp)
             self.history.append(temp)
+            self.push(temp)
```
What was the significance of this, why is it important to push append to history first?

rybarczykj replied:
Oh, right. It makes no difference the order of those two.

But, I meant to also append to all_logical_lines right there, which _does_ have to be before self.push(temp). Rearranging the order of self.history would keep those three steps in the same order as is in on_enter.

Adding that commit now to add all_logical_lines there.

thomasballinger replied:
cool, seems fine to reorder for consistency

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -1847,12 +1837,20 @@ def take_back_buffer_line(self):
-        self.history.pop()
+        undone_input = self.history.pop()
+
+        # Pop all outputs leading up to the undone input
```
We shouldn't need this for `take_back_buffer_line()`, the buffer is the current (as yet unexecuted) possibly-multiline input statement, so removing a line will not affect output

thomasballinger commented on bpython/curtsiesfrontend/repl.py:
```diff
@@ -408,13 +409,16 @@ def __init__(
+        
+        # This is to use "input" or "output" for all_logical_lines tuples without actually 
+        # having to know the string.
+        class lineTypeTranslator(Enum):
```
I'd move this out to the global scope so that other code can access it without needing to use self.lifeTypeTranslator, and use TitleCase instead of camelCase

rybarczykj replied:
Got it. I added it as a class in bpython.repl, and now curtsiesfrontend.repl imports it.
